perf_benchmark/benchmark_main.cc

```diff
@@ -0,0 +1,497 @@
+          + actual_proto.DebugString()).c_str());
+    }
+  }
+}
+
+static void BM_Int32ArrayTypePayloadFromJsonNonStreaming(::benchmark::State& state) {
+  BM_ArrayPayloadFromJson<Int32ArrayPayload>(state,
+                                             kInt32ArrayPayloadMessageType,
+                                             false,
+                                             0);
+}
+static void BM_DoubleArrayTypePayloadFromJsonNonStreaming(::benchmark::State& state) {
+  BM_ArrayPayloadFromJson<DoubleArrayPayload>(state,
+                                              kDoubleArrayPayloadMessageType,
+                                              false,
+                                              0);
+}
+static void BM_StringArrayTypePayloadFromJsonNonStreaming(::benchmark::State& state) {
+  BM_ArrayPayloadFromJson<StringArrayPayload>(state,
+                                              kStringArrayPayloadMessageType,
+                                              false,
+                                              0);
+}
+BENCHMARK_WITH_PERCENTILE(BM_Int32ArrayTypePayloadFromJsonNonStreaming);
+BENCHMARK_WITH_PERCENTILE(BM_DoubleArrayTypePayloadFromJsonNonStreaming);
+BENCHMARK_WITH_PERCENTILE(BM_StringArrayTypePayloadFromJsonNonStreaming);
+
+//
+// Benchmark variable: Number of nested JSON layer.
+//
+// Helper function for benchmarking translation from nested JSON values.
+void BM_NestedPayloadFromJson(::benchmark::State& state,
+                              uint64_t layers,
+                              bool streaming,
+                              uint64_t stream_size,
+                              absl::string_view msg_type) {
+  auto json_msg =
+      absl::make_unique<std::string>(GetNestedJsonString(layers,
+                                                         kNestedFieldName,
+                                                         std::string(
+                                                             kInnerMostNestedFieldName),
+                                                         "buzz"));
+  absl::StatusOr<std::string> proto_str = BenchmarkJsonTranslation(state,
+                                                                   msg_type,
+                                                                   json_msg.get(),
+                                                                   streaming,
+                                                                   stream_size,
+                                                                   1);
+  if (!proto_str.ok()) {
+    state.SkipWithError(proto_str.status().ToString().c_str());
+  } else {
+    // Verification
+    int actual_layers = 0;
+    if (msg_type == kNestedPayloadMessageType) {
+      NestedPayload actual_proto;
+      actual_proto.ParseFromString(*proto_str);
+
+      NestedPayload* it = &actual_proto;
+      while (it->has_nested()) {
+        ++actual_layers;
+        it = it->mutable_nested();
```

Comment:
why use "mutable_"?

Reply:
Great question. Thanks for bringing this up! I didn't have too much experience with protobuf in c++. After looking into it, I should let the pointer be `const NestedPayload*` instead of just `NestedPayload*`, then I can use `it->nested()` instead of `it->mutable_nested()`.